CRM/ManualDirectDebit/Batch/BatchHandler.php

```diff
@@ -0,0 +1,148 @@
+    }
+
+    return $this->batchStatusId;
+  }
+
+  /**
+   * Gets batch status
+   *
+   * @return string
+   */
+  public function getBatchStatus() {
+    if (!$this->batchStatus) {
+      $statusID = $this->getBatchStatusId();
+      $batchStatuses = CRM_Core_PseudoConstant::get('CRM_Batch_DAO_Batch', 'status_id', [
+        'labelColumn' => 'name',
+        'status' => " v.value={$statusID}",
+      ]);
+      $this->batchStatus = $batchStatuses[$statusID];
+    }
+
+    return $this->batchStatus;
+  }
+
+  /**
+   * Checks whether the batch is open
+   *
+   * @return bool
+   */
+  public function validBatchStatus() {
+    return in_array($this->getBatchStatus(), ['Open', 'Reopened']);
+  }
+
+  /**
+   * Creates Export File
+   *
+   * @param array $params
+   */
+  public function createExportFile($params) {
+    $returnValues = [
+      'contact_id' => CRM_ManualDirectDebit_Batch_Transaction::DD_MANDATE_TABLE . '.entity_id as contact_id',
+      'name' => CRM_ManualDirectDebit_Batch_Transaction::DD_MANDATE_TABLE . '.account_holder_name as name',
+      'sort_code' => CRM_ManualDirectDebit_Batch_Transaction::DD_MANDATE_TABLE . '.sort_code as sort_code',
+      'account_number' => CRM_ManualDirectDebit_Batch_Transaction::DD_MANDATE_TABLE . '.ac_number as account_number',
+      'amount' => 'IF(civicrm_contribution.net_amount IS NOT NULL, civicrm_contribution.net_amount , 0) as amount',
+      'reference_number' => CRM_ManualDirectDebit_Batch_Transaction::DD_MANDATE_TABLE . '.dd_ref as reference_number',
+      'transaction_type' => 'civicrm_option_value.label as transaction_type',
+    ];
+
+    $batchTransaction = new CRM_ManualDirectDebit_Batch_Transaction($params['batch_id'], $params, [], $returnValues);
+
+    $mandateItems = $batchTransaction->getDDMandateInstructions();
+
+    $headers = [
+      ts('Contact ID'),
+      ts('Account Holder Name'),
+      ts('Sort code'),
+      ts('Account Number'),
+      ts('Amount'),
+      ts('Reference Number'),
+      ts('Transaction Type'),
+    ];
+
+    $contents = $this->generateCSVFile($headers, $mandateItems);
+
+    $fileName = 'Batch_' . $this->batchID . '_' . date('YmdHis') . '.csv';
+    CRM_Utils_System::setHttpHeader('Content-Type', 'text/plain');
+    CRM_Utils_System::setHttpHeader('Content-Disposition', 'attachment; filename=' . CRM_Utils_File::cleanFileName(basename($fileName)));
+    ob_clean();
+    flush();
+    echo $contents;
+    CRM_Utils_System::civiExit();
+  }
+
+  /**
+   * Creates csv File
+   *
+   * @param array $headers
+   *
+   * @param \CRM_Core_DAO $export
+   *
+   * @return string
+   */
+  private function generateCSVFile($headers, $export) {
+    $out = fopen('php://temp/maxmemory:'. (12*1024*1024), 'r+');
+    fputcsv($out, $headers);
+
+    while ($export->fetch()) {
+      fputcsv($out, $export->toArray());
+    }
+    rewind($out);
+    $contents = stream_get_contents($out);
+    fclose($out);
+
+    return $contents;
```

Comment:
The problem here is that you're still loading the whole content of the file to memory when you assign the contents of the stream to `$contents`. To avoid this, we need to send the contents of the stream directly to output, by using `fpssthrough($out)`, or echoing line by line instead. Maybe you could return the rewinded file descriptor? Or assign the file descriptor to a class attribute, whichever makes more sense to you.

Eg.
```php
public function createExportFile($params) {
  ...
  ob_clean();
  flush();
  $csvFilePointer = $this->generateCSVFile($headers, $mandateItems);
  fpassthrough($csvFilePinter);
  fclose($csvFilePointer);
  CRM_Utils_System::civiExit();
}

private function generateCSVFile($headers, $export) {
  ...
  rewind($out);
  
  return $out;
}
```

Or if you find it too weird to return open file pointers, maybe we could just do everything inside generateCSVFile, so the pointer is not returned, and the contents of the file are immediately passed through to output.

```php
public function createExportFile($params) {
    CRM_Utils_System::setHttpHeader('Content-Type', 'text/plain');
    CRM_Utils_System::setHttpHeader('Content-Disposition', 'attachment; filename=' . 
    CRM_Utils_File::cleanFileName(basename($fileName)));
    ob_clean();
    flush();
    $this->generateCSVFile($headers, $mandateItems); // Although I'd change the name of this function to outputCSVFile() if I chose this approach.
    CRM_Utils_System::civiExit();
}
```